### DISCUSSION TITLE
pnpm support?

### DISCUSSION THREAD

**Discussion by aozaki-kuro** [vuepress/core#695] [Q&A] [answered]
Just heard the news (actually in Oct 2020) that Vue, Vite and VitePress are all using pnpm instead of yarn v1 classic now.

Vue: https://github.com/vuejs/core/pull/4766
Vite: https://github.com/vitejs/vite/pull/5060

VuePress still recommends npm or yarn while pnpm needs a `shamefully-hoist=true` hack to have it running.

pnpm is really awesome in saving time and disk storage so I am looking forward to a native support... Anyone?

**Comment by Mister-Hope** [comment]
Nope, unless we fully reshape our api.

Currently api from v1 are depending on filepath heavily. So it won't work if there is acually no such files under `node_modules`, like `clientEnhanceAppFiles`.

Also, we had to drop alia support, like `theme:'hope'` for using `vuepress-theme-hope`, so does the plugins, as the `@vuepress/core` and `@vuepress/cli` are not actually depending on those deps.

We have to change theme plugins to let them export a function, also use module path in all paths:

Before:

```ts
// .vuepress/config.ts
export default {
  theme: 'hope',
  themeConfig: {
    // config
  },
  plugins: [
    ['blog2', { 
      // options
    }],
}

// plugin api in vuepress-plugin-blog2 under lib/node
import { path } from '@vuepress/utils'

export default (options) => {
   clientAppEnhanceFiles: path.resolve(__dirname, '../client/appEnhance.js')
}
```


After:

```ts
// .vuepress/config.ts
import { useTheme } from 'vuepress-theme-hope';
import { useBlog2 } from 'vuepress-theme-blog2';

export default {
  theme: useTheme({
    // themeConfig
  }),
  plugins: [
    useBlog2({ 
      // options
    }),
  ]
}

// plugin api in vuepress-plugin-blog2 under lib/node
export default (options) => {
   clientAppEnhanceFiles: 'vuepress-plugin-blog/lib/client/appEnhance.js'
}

// package.json for vuepress-plugin-blog2

{
  ...
  exports: {
    './lib/client/appEnhance.js': './lib/client/appEnhance.js'
  },
 ...
}
```

That's a HUGE change.

Also only the changes in plugin api is necssary for `hoist`, but a full support for Plug'inPlay needs the first one.

I once started a dicussion before, but @meteorlxy  refuse it.

---

BTW: I eager this feature as well, and I am already doing the above changes to make my theme and plugins ready. So maybe you can try to convince @meteorlxy  that this change is needed.

**Comment by Mister-Hope** [comment]
It's not a good time to start this huge change after comming to `beta.35` after all. Is my description why we are not supporting it clear enough to you?

**Reply by aozaki-kuro** [reply]
Welp seemed like it's a real big project once started. It's still usable now so maybe I would stick with the present hack for a while 😕

**Reply by Mister-Hope** [reply]
I hate the current method as well because spaces and install time, but the final decision could only be made by meteorlxy, and he seems not interested in it the last time I discussed with him.

If he change his mind, you will get noticed through this dicussion.

**Reply by aozaki-kuro** [reply]
Honestly I'm considering moving to VitePress as well since nothing except dark mode is irreplaceable for now...
But I would like to wait and see  what would this project be like in the future. Best wishes on development 🙏

**Comment by meteorlxy** [comment]
UPDATE:

Good news, we have moved to pnpm after a huge refactor since 2.0.0-beta.40 (with lots of breaking changes though 😅 )

-----

LEGACY ANSWER:

It is inappropriate to compare vuepress with vue/vite. Also, even the vue/vite repo move to use pnpm, it does not means that users should also use pnpm, and vice versa.

A similar project is [slidev](https://sli.dev/guide/install.html#install-manually), which also loads themes dynamically from `node_modules`, and also requires `shamefully-hoist=true`.

In brief, pnpm limits a project to load dependencies dynamically, so we have to make pnpm's behavior consistent with npm / yarn v1 to work properly.

A [.pnpmfile.cjs](https://pnpm.io/pnpmfile) file in your project could help, but not such easy.

---

As for VitePress, it uses a different way to load themes (like [this](https://github.com/vuejs/docs/blob/34f37ade9b7d5c84cdde6da4d3c7c21e053c22db/.vitepress/config.ts#L4-L6) and [this](https://github.com/vuejs/docs/blob/34f37ade9b7d5c84cdde6da4d3c7c21e053c22db/.vitepress/theme/index.ts#L3)), i.e. explictly install and import the theme package.

It's quite different from VuePress v1 /v2 way. We could change to use that way, which will cause breaking changes for sure.

In addition, VitePress also does not support vitepress-specific plugins, and users need to install and use vite plugin directly. You know, current VuePress plugin mechanism is also quite different 😞 .

**Comment by meteorlxy** [comment, marked as answer]
Good news, we have moved to pnpm after a huge refactor (with lots of breaking changes though 😅 )

**Reply by aozaki-kuro** [reply]
Good to know that 💪
Would try to migrate my existing VuePress repo later!
